modules/47_regipol/regiCarbonPrice/datainput.gms

```diff
@@ -26,13 +26,20 @@ p47_nonEnergyUse("2030",ext_regi)$(sameas(ext_regi, "EU27_regi")) = 0.11;
 *** Emission markets (EU Emission trading system and Effort Sharing)
 ***--------------------------------------------------
 $IFTHEN.emiMkt not "%cm_emiMktTarget%" == "off" 
-  
+
 *** Auxiliar parameters based on emission targets information 
   loop((ttot,ttot2,ext_regi,emiMktExt,target_type_47,emi_type_47)$pm_emiMktTarget(ttot,ttot2,ext_regi,emiMktExt,target_type_47,emi_type_47), !!calculated sets that depends on data parameter
     regiEmiMktTarget(ext_regi) = yes;
     regiANDperiodEmiMktTarget_47(ttot2,ext_regi) = yes;
   );
 
+  loop((ext_regi,ttot,ttot2,emiMktExt,target_type_47,emi_type_47)$pm_emiMktTarget(ttot,ttot2,ext_regi,emiMktExt,target_type_47,emi_type_47), !!calculated sets that depends on data parameter
+    loop(regi$regi_groupExt(ext_regi,regi),
+      regiEmiMktTarget2regi_47(ext_regi2,regi) = NO;
+      regiEmiMktTarget2regi_47(ext_regi,regi) = YES;
```

Comment:
Let me try to explain with an example:

We want to set a emission target at EU level, and another one at German level. 
To do that we can set `cm_emiMktTarget` to `2020.2050.EU27_regi.all.year.netGHG 0.01, 2020.2045.DEU.all.year.netGHG 0.01`

In the code, we want the `DEU` target to be responsible to determine the German carbon tax, and the `EU27_regi` target to set the price for all other EU regions, except for Germany.

This code is designed to create a set tupple that reflects that. 
So the result in this example would be to create a set like the one bellow:
```
set regiEmiMktTarget2regi_47(ext_regi,regi) /
 EU27 (ENC,EWN,ECS,ESC,ECE,FRA,ESW)  !!EU27 without DEU because the carbon price for it should be defined by the DEU target  
 DEU DEU
/;
```
in which `ext_regi` corresponds to the dimension that you defined your emission targets, and `regi` corresponds to the regions that have their carbon price controlled by it.
 

Now lets talk about how the code above achieves that:

Firstly, we loop through all possible emissions targets set at `cm_emiMktTarget`.
However, pay attention that in the loop I am looping first in the dimension `ext_regi`. I am doing this to follow an specific order of evaluation, going through the most aggregated regions to the less aggregated ones.

Any statement in GAMs is executed following the order defined at the `GAMS entry order` list. 
As the set `ext_regi` is defined from the biggest level of regional aggregation (`GLO`), to the smallest one (model region `USA`, `DEU`, ...), as long as no item of this set is previously used in another set in the REMIND code, the `GAMS entry order` will respect the order I declared in this set.
As consequence, if I loop over `ext_regi`, we will first runt the code for `GLO`, then the aggregated regions (e.g. `EUR_regi`, `EU27_regi`, ...), and only at last the more disaggregated regions that correspond to the regions actually used in the optimization code (`USA`, `DEU`, and so on).

So, back to our example, we are going to firstly run the loop for `EU27_regi` and only after to `DEU`.

The inner loop (`loop(regi$regi_groupExt(ext_regi,regi),`) is responsible to translate `ext_regi` into the model used regions (`regi`). So when you are running the code for `EU27_regi`, you are actually running it for every region contained in `EU27_regi` (`ENC,EWN,ECS,ESC,ECE,FRA,ESW,DEU`). When you run the code for a model region, e.g. `DEU`, you will only run it for `DEU`.

As we are executing in the example above the aggregated region first, when we run `EU27_regi` the statement `regiEmiMktTarget2regi_47(ext_regi2,regi) = NO;` will do nothing, but the second statement `regiEmiMktTarget2regi_47(ext_regi,regi) = YES;` will include all regions inside `EU27_regi` to the calculated set.
The result after running only the `EU27_regi` will be:

```
set regiEmiMktTarget2regi_47(ext_regi,regi) /
 EU27 (ENC,EWN,ECS,ESC,ECE,FRA,ESW,DEU)
/;
```

But as we said before, we don't want to have `DEU` carbon prices controlled by the `EU27_regi` target. 

So let's keep going in the loop.

The next item in the loop will be to run it for the non aggregated region `DEU`.

When we run the same code for this region, the first code line will remove `DEU` from any other possible item in the set, in this case remove  `DEU` from the `EU27_regi` item above, meanwhile the second statement will add a new entry `DEU DEU`, to reflect that targets that use DEU should control the German carbon price.

As as result, we will have:
```
set regiEmiMktTarget2regi_47(ext_regi,regi) /
 EU27 (ENC,EWN,ECS,ESC,ECE,FRA,ESW)  !!EU27 without DEU because the carbon price for it should be defined by the DEU target  
 DEU DEU
/;
```
which is exactly what we wanted in the first place.

Now in the carbon price update code I have a set that determines exactly which regions should have their carbon prices updated for a given target, without intersections anymore as it was happening before.

Reply:
wow, impressive, and thanks for the detailed explanation!

a) I guess this would be a perfect section for a description/tutorial page for regipol. Do we already have such a document for regipol where we can first compile such information snippets and then later bring it into a nice structure?

b) I think a line mentioning at the beginning of these 8 lines code that the following code makes sure only the smallest (in a regional sense) emission target applicable to a region is applied in the model, and that it relies on the set  `ext_regi` to always be defined in such a way that subsets of regions are defined AFTER their supersets would be helpful in preventing future modelers to not understand it/destroy it by accident. 
Similarly, such a heads-up in core/sets would be good :-)

Reply:
and one needs to keep in mind that targets for subset of regions should be tighter than the overall target. 
So if we set the ECEMF USA runs, we should NOT add an US target as it would be weaker than the global target, and that would (at least in my understanding) not be the aim of the protocol.

Reply:
Nice idea of having a more detailed description/tutorial section, but I am not sure the format that I should follow for that. I should get in contact with RSE to clarify that first.

I will also go through your suggestions for better comments later in the afternoon so I can add them to this PR.

Reply:
> and one needs to keep in mind that targets for subset of regions should be tighter than the overall target. So if we set the ECEMF USA runs, we should NOT add an US target as it would be weaker than the global target, and that would (at least in my understanding) not be the aim of the protocol.

You can actually add a less stringent target for an specific region like USA. As long as other regions have a way to compensate that to not create infeasibilities, the results will reflect that your USA carbon prices are lower than the other regions carbon pricing, which could be the desired effect for some specific scenario.

Reply:
addressed at https://github.com/remindmodel/remind/pull/1315/commits/cb7953ffda7c42ac9699f3aeb77b95f0f790ede3